engine/crates/integration-tests/src/federation/request.rs

```diff
@@ -0,0 +1,160 @@
+mod stream;
+
+use std::{
+    borrow::Cow,
+    future::IntoFuture,
+    ops::{Deref, DerefMut},
+    str::FromStr,
+    sync::Arc,
+};
+
+use engine::{BatchRequest, Variables};
+use engine_v2::{HttpGraphqlResponse, HttpGraphqlResponseBody};
+use futures::future::BoxFuture;
+use http::{header::Entry, HeaderName, HeaderValue};
+use serde::de::Error;
+pub use stream::*;
+
+use crate::engine_v1::GraphQlRequest;
+
+use super::TestRuntime;
+
+#[must_use]
+pub struct ExecutionRequest {
+    pub(super) request: GraphQlRequest,
+    #[allow(dead_code)]
+    pub(super) headers: Vec<(String, String)>,
+    pub(super) engine: Arc<engine_v2::Engine<TestRuntime>>,
+}
+
+impl ExecutionRequest {
+    pub fn by_client(self, name: &'static str, version: &'static str) -> Self {
+        self.header("x-grafbase-client-name", name)
+            .header("x-grafbase-client-version", version)
+    }
+
+    /// Adds a header into the request
+    pub fn header(mut self, name: impl Into<String>, value: impl Into<String>) -> Self {
+        self.headers.push((name.into(), value.into()));
+        self
+    }
+
+    pub fn variables(mut self, variables: impl serde::Serialize) -> Self {
+        self.request.variables = Some(Variables::from_json(
+            serde_json::to_value(variables).expect("variables to be serializable"),
+        ));
+        self
+    }
+
+    pub fn extensions(mut self, extensions: impl serde::Serialize) -> Self {
+        self.request.extensions =
+            serde_json::from_value(serde_json::to_value(extensions).expect("extensions to be serializable"))
+                .expect("extensions to be deserializable");
+        self
+    }
+
+    fn http_headers(&self) -> http::HeaderMap {
+        let mut headers = http::HeaderMap::new();
+
+        for (key, value) in &self.headers {
+            let key = HeaderName::from_str(key).unwrap();
```

Comment:
:zap: **Suggestion** 
Consider handling potential errors when using `unwrap()` to avoid panics. Specifically, `unwrap()` is used on `HeaderName::from_str`, `HeaderValue::from_str`, `Box::pin`, and `serde_json::to_string_pretty`. In each of these cases, if the operation fails, it could lead to a panic. Handling these errors more gracefully would improve the robustness of the code.